src/Microsoft.AspNetCore.Mvc.Razor/Compilation/MetadataReferenceFeatureProvider.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection.PortableExecutable;
+using Microsoft.AspNetCore.Mvc.ApplicationParts;
+using Microsoft.CodeAnalysis;
+using Microsoft.Extensions.DependencyModel;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Compilation
+{
+    /// <summary>
+    /// An <see cref="IApplicationFeatureProvider{TFeature}"/> for <see cref="MetadataReferenceFeature"/> that 
+    /// uses <see cref="DependencyContext"/> for registered <see cref="AssemblyPart"/> instances to create 
+    /// <see cref="MetadataReference"/>.
+    /// </summary>
+    public class MetadataReferenceFeatureProvider : IApplicationFeatureProvider<MetadataReferenceFeature>
+    {
+        /// <inheritdoc />
+        public void PopulateFeature(IEnumerable<ApplicationPart> parts, MetadataReferenceFeature feature)
+        {
+            if (parts == null)
+            {
+                throw new ArgumentNullException(nameof(parts));
+            }
+
+            if (feature == null)
+            {
+                throw new ArgumentNullException(nameof(feature));
+            }
+
+            var libraryPaths = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            foreach (var part in parts.OfType<AssemblyPart>())
+            {
+                var dependencyContext = DependencyContext.Load(part.Assembly);
+                if (dependencyContext == null)
+                {
+                    continue;
+                }
+
+                for (var i = 0; i < dependencyContext.CompileLibraries.Count; i++)
+                {
+                    var library = dependencyContext.CompileLibraries[i];
```

Comment:
Why do we use the CompileLibraries here instead of the RuntimeLibraries as in https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs#L66?